blog/_posts/2022-12-23-wildcart-domain-with-caddy-for-tunnel-hosting.md

```diff
@@ -0,0 +1,137 @@
+
+While using my tunnel for local development and testing I found that the services a want to expose regularly change. It can become inconvenient to configure many domains and keep DNS records up to date.
+
+By using a wildcard domain for the tunnel it would be possible to expose services on multiple subdomains without having to configure the tunnel server and DNS for each subdomain.
+
+Let's Encrypt requires the [DNS challenge](https://letsencrypt.org/docs/challenge-types/#dns-01-challenge) to obtain wildcard certificates. This is not supported by the inlets. To make it work we will show you how to use [Caddy](https://caddyserver.com/) as a reverse proxy for the inlets HTTP tunnel. Caddy will take care of certificate renewals and TLS termination.
+
+In the next sections we will walk you through through the setup and configuration of inlets and caddy.
+
+## Pre-reqs
+
+- A Linux server with a public IP.
+- An [inlets license](https://inlets.dev/pricing/)
+- Access to the dashboard or API of your domain provider to setup the DNS challenge and DNS records.
+
+## Run the inlets http tunnel server 
+
+Download the inlets-pro binary:
+ 
+ - Download it from the [GitHub releases](https://github.com/inlets/inlets-pro/releases)
+ - Get it with [arkade](https://github.com/alexellis/arkade): `arkade get inlets-pro`
+
+Generate an authentication token the client will use to connect to the server:
+```bash
+TOKEN="$(head -c 32 /dev/urandom | base64 | cut -d "-" -f1)"
+```
+
+Configure the inlets tunnel server. The `--generate` flag can be used to turn the command arguments into a configuration file for systemd. The usual place to save the service file is: `/etc/systemd/system/inlets.service`.
+
+```bash
+inlets-pro http server \
+  --auto-tls false \
+  --control-addr 127.0.0.1 \
+  --data-addr 127.0.0.1 \
+  --token $TOKEN \
+  --generate=systemd > inlets.service
+```
+
+Notice that we are binding the control plane and data plane to the local host only. Auto TLS is also disabled. In the next section we will setup caddy as a reverse proxy to make the tunnel accessible on the public internet and handle TLS termination.
+
+Move the service file and start the service:
+
+```bash
+sudo mv inlets.service /etc/systemd/system/inlets.service
+sudo systemctl daemon-reload
+sudo systemctl enable --now inlets
+```
+
+Verify that the service is running:
+
+```bash
+systemctl status inlets
+```
+
+## Setup Caddy
+
+Caddy will be used as a reverse proxy for the inlets http tunnel server. It will   
+
+By default the Caddy server does not contain any DNS modules. These can be added to your download from [caddyserver.com](https://caddyserver.com/download). Alternatively you can use xcaddy to build caddy with your DNS provider plugged in. The [caddy community wiki](https://caddy.community/t/how-to-use-dns-provider-modules-in-caddy-2/8148) has more details on how to enable the DNS challenge for your provider.
```

Comment:
Do you want to say which one you used for Cloudflare? And which you might need for DigitalOcean who also have DNS01?